### PR TITLE
fix: format addresses according to a thread-local network flag

### LINKED ISSUE #3018: Get rid of `current_network` global state
**Issue summary**

The FVM tracks which network is currently active: Either a mainnet or a testnet. If mainnet is active, only mainnet address may be parsed (they are prefixed with `f`). If a testnet is active, only testnet address may be parsed (they are prefixed with `t`).

Using global state doesn't work when we want to deal with both mainnet address and testnet addresses in different threads simultaneously. We need to consider our options for moving away from global state entirely.

Possible options:
 - Remove the global state and ignore the address prefixes. This would mean we don't show an error if someone tries to use a testnet address on the mainnet.
 - Remove the global state and manually check that the addresses are prefixed correctly. This involves more work.

**Other information and links**

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- Relax parsing of the `Address` structure. It now accepts both mainnet and testnet addresses.
- Introduce `StrictAddress` which only parses the addresses of the current network.
- Change the `current_network` state to be thread-local.

## Reference issue to close (if applicable)

Closes #3018 

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Review by LesnyRumcajs — APPROVED** [review]
:smiling_face_with_tear:

**Review by ruseinov — CHANGES_REQUESTED** [review]
nit: let's create an issue to fix this properly.
another nit: let's modify CI to run the ignored tests via nextest instead of just disabling them for good. This would perhaps protect us from introducing any regressions while we're coming up with a proper fix.

**Comment by lemmih** [comment]
> nit: let's create an issue to fix this properly.

Here's the issue: https://github.com/ChainSafe/forest/issues/3018

**Review by elmattic — APPROVED** [review]

File: src/cli/main.rs [hunk 1]
```diff
@@ -35,7 +35,7 @@ pub async fn main() -> anyhow::Result<()> {
                 }
             }
             if config.chain.is_testnet() {
-                crate::shim::address::set_current_network(crate::shim::address::Network::Testnet);
+                crate::shim::address::CurrentNetwork::set_global(crate::shim::address::Network::Testnet);
```
Comment (ruseinov) [inline src/cli/main.rs:38]: Seems like a good use-case for importing `crate::shim::address::{..}`.

File: src/daemon/daemon.rs [hunk 2]
```diff
@@ -111,7 +111,7 @@ pub(super) async fn start(
     shutdown_send: mpsc::Sender<()>,
 ) -> anyhow::Result<()> {
     if config.chain.is_testnet() {
-        crate::shim::address::set_current_network(crate::shim::address::Network::Testnet);
+        crate::shim::address::set_global_network(crate::shim::address::Network::Testnet);
```
Comment (ruseinov) [inline src/daemon/daemon.rs:114]: I think this is intended to be `crate::shim::address::CurrentNetwork::set_global`.

**Comment by ruseinov** [comment]
Looking at the `thread_local` code it feels like things are becoming a bit convoluted with this global variable and it's handling. 

Is there a reason why we need a global variable at all in `fvm` and as a consequence in `forest`? It seems to be that this design might be coming from the `lotus` implementation. I'm not sure if this is a good place to discuss this, but would be nice to take a look at that going forward to see if there are better ways.

File: src/daemon/daemon.rs (hunk 2 above)
Comment (lemmih) [reply src/daemon/daemon.rs:114]: This file is an orphan and will be deleted. I didn't even mean to edit it.

File: src/shim/address.rs [hunk 3]
```diff
@@ -25,6 +26,65 @@ lazy_static! {
...
 }
 
+static GLOBAL_NETWORK: AtomicU8 = AtomicU8::new(Network::Mainnet as u8);
+
+thread_local! {
+    // Thread local network identifier. Defaults to value in GLOBAL_NETWORK.
+    static LOCAL_NETWORK: AtomicU8 = AtomicU8::new(GLOBAL_NETWORK.load(Ordering::Relaxed));
+}
+
+// For user safety, Filecoin has different addresses for its mainnet and test networks: Mainnet
```
[the remark this answers is not in the document]
Comment (lemmih) [reply src/shim/address.rs]: Done.

**Comment by lemmih** [comment]
> Looking at the `thread_local` code it feels like things are becoming a bit convoluted with this global variable and it's handling.
> 
> Is there a reason why we need a global variable at all in `fvm` and as a consequence in `forest`? It seems to be that this design might be coming from the `lotus` implementation. I'm not sure if this is a good place to discuss this, but would be nice to take a look at that going forward to see if there are better ways.

The alternative is to pass around an explicit `Network` parameter almost everywhere. It would be a significant change and it might not be worth it.

File: src/shim/address.rs [hunk 4]
```diff
@@ -86,15 +156,79 @@ impl FromStr for Address {
...
+        // write `fP` where P is the protocol number.
+        write!(f, "{}{}", prefix, protocol)?;
+
+        fn write_payload(
+            f: &mut std::fmt::Formatter<'_>,
+            protocol: Protocol,
+            prefix: Option<&[u8]>,
+            data: &[u8],
+        ) -> std::fmt::Result {
+            let mut hasher = blake2b_simd::Params::new()
```
[the remark this answers is not in the document]
Comment (lemmih) [reply src/shim/address.rs:197]: Is there an alternative?

[the remark this answers is not in the document]
Comment (ruseinov) [reply src/shim/address.rs:197]: Looking into it.

[the remark this answers is not in the document]
Comment (ruseinov) [reply src/shim/address.rs:197]: My question would be as follows: do we actually need all this for display? 

And if so - why? 

If this is just needed to check in tests - we could get away with some sort of hex representation as long as we can differentiate from one another, making sure that unique payload yields a unique string representation.

**Review by ruseinov — APPROVED** [review]

**Comment by ruseinov** [comment]
I am happy with this now that I've taken the time to re-read it.

File: src/shim/address.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lemmih) [reply src/shim/address.rs:197]: The format is defined in the [specification](https://spec.filecoin.io/appendix/address/). The last 4 bytes have to be the blake2b hash of the payload.

**Pull request merged**